ktor-server/ktor-server-core/jvm/src/io/ktor/application/newapi/KtorPlugin.kt

```diff
@@ -0,0 +1,655 @@
+                insertToPhaseRelatively(
+                    currentPlugin.beforeResponseInterceptions,
+                    otherPlugin.beforeResponseInterceptions,
+                    block
+                )
+            }
+
+            override fun after(block: suspend ResponseExecution.(ApplicationCall) -> Unit) {
+                insertToPhaseRelatively(
+                    currentPlugin.afterResponseInterceptions,
+                    otherPlugin.afterResponseInterceptions,
+                    block
+                )
+            }
+        }
+    }
+
+    public class AfterPluginContext(currentPlugin: InterceptionsHolder, otherPlugin: InterceptionsHolder) :
+        RelativePluginContext(currentPlugin, otherPlugin) {
+        override fun selectPhase(phases: List<PipelinePhase>): PipelinePhase? = phases.lastOrNull()
+
+        override fun insertPhase(
+            pipeline: Pipeline<*, ApplicationCall>,
+            relativePhase: PipelinePhase,
+            newPhase: PipelinePhase
+        ) {
+            pipeline.insertPhaseAfter(relativePhase, newPhase)
+        }
+    }
+
+    public class BeforePluginContext(currentPlugin: InterceptionsHolder, otherPlugin: InterceptionsHolder) :
+        RelativePluginContext(currentPlugin, otherPlugin) {
+        override fun selectPhase(phases: List<PipelinePhase>): PipelinePhase? = phases.firstOrNull()
+
+        override fun insertPhase(
+            pipeline: Pipeline<*, ApplicationCall>,
+            relativePhase: PipelinePhase,
+            newPhase: PipelinePhase
+        ) {
+            pipeline.insertPhaseBefore(relativePhase, newPhase)
+        }
+    }
+
+    /**
+     * Execute some actions right after some other [plugin] was already executed.
+     *
+     * Note: you can define multiple actions inside a [build] callback for multiple stages of handling an HTTP call
+     * (such as [onCall], [onResponse], etc.) and each of these actions will be executed right after all actions defined
+     * by the given [plugin] were already executed in the same stage.
+     * */
+    public fun afterPlugin(plugin: InterceptionsHolder, build: AfterPluginContext.() -> Unit): Unit =
+        AfterPluginContext(this, plugin).build()
+
+    /**
+     * Execute some actions right before some other [plugin] was already executed.
+     *
+     * Note: you can define multiple actions inside a [build] callback for multiple stages of handling an HTTP call
+     * (such as [onCall], [onResponse], etc.) and each of these actions will be executed right before all actions defined
+     * by the given [plugin] were already executed in the same stage.
+     * */
+    public fun beforePlugin(plugin: InterceptionsHolder, build: BeforePluginContext.() -> Unit): Unit =
+        BeforePluginContext(this, plugin).build()
+
+    public companion object {
+        /**
+         * A canonical way to create a [KtorPlugin].
+         * */
+        public fun <Configuration : Any> createPlugin(
```

Comment:
**Important**
The big problem here is that `createPlugin` returns instance of a plugin, while it should return factory for a plugin. 
It could be an issue if the feature has a private state and, after installing it on different scopes, this state will be shared. Imagine you have a plugin that counts how much time some routing subtree was called:  
```
val CoutingPlugin = createPlugin { 
  var counter = ...
  onCall { counter++ }
}
```
And if you install it in different subtrees like
```
route("a") {
  install(CoutingPlugin) 
  handle { ... }
}
route("b") {
  install(CoutingPlugin) 
  handle { ... }
}
```
`counter` field will be shared between these two routes.